pkg/progress/command.go

```diff
@@ -48,92 +35,37 @@ func RunOSBuild(pb ProgressBar, manifest []byte, exports []string, opts *OSBuild
-	}
-	cmd := exec.Command(
-		osbuildCmd,
-		"--store", opts.StoreDir,
-		"--output-directory", opts.OutputDir,
-		fmt.Sprintf("--cache-max-size=%v", cacheMaxSize),
-		"-",
-	)
-	for _, export := range exports {
-		cmd.Args = append(cmd.Args, "--export", export)
-	}
-	cmd.Env = append(os.Environ(), opts.ExtraEnv...)
-	cmd.Stdin = bytes.NewBuffer(manifest)
-	return cmd
-}
-
-func runOSBuildNoProgress(pb ProgressBar, manifest []byte, exports []string, opts *OSBuildOptions) error {
-	var stdout, stderr io.Writer
-
-	var writeMu sync.Mutex
-	if opts.BuildLog == nil {
-		// No external build log requested and we won't need an
-		// internal one because all output goes directly to
-		// stdout/stderr. This is for maximum compatibility with
-		// the existing bootc-image-builder in "verbose" mode
-		// where stdout, stderr come directly from osbuild.
-		stdout = osStdout()
-		stderr = osStderr()
-	} else {
-		// There is a slight wrinkle here: when requesting a
-		// buildlog we can no longer write to separate
-		// stdout/stderr streams without being racy and give
-		// potential out-of-order output (which is very bad
-		// and confusing in a log). The reason is that if
-		// cmd.Std{out,err} are different "go" will start two
-		// go-routine to monitor/copy those are racy when both
-		// stdout,stderr output happens close together
-		// (TestRunOSBuildWithBuildlog demos that). We cannot
-		// have our cake and eat it so here we need to combine
-		// osbuilds stderr into our stdout.
-		mw := newSyncedWriter(&writeMu, io.MultiWriter(osStdout(), opts.BuildLog))
-		stdout = mw
-		stderr = mw
-	}
-
-	cmd := newOsbuildCmd(manifest, exports, opts)
-	cmd.Stdout = stdout
-	cmd.Stderr = stderr
+func runOSBuildNoProgress(pb ProgressBar, manifest []byte, exports []string, opts *osbuild.OSBuildOptions) error {
+	cmd := osbuild.NewOSBuildCmd(manifest, exports, opts)
 	if err := cmd.Run(); err != nil {
 		return fmt.Errorf("error running osbuild: %w", err)
 	}
 	return nil
 }
 
-var osbuildCmd = "osbuild"
-
-func runOSBuildWithProgress(pb ProgressBar, manifest []byte, exports []string, opts *OSBuildOptions) (err error) {
+func runOSBuildWithProgress(pb ProgressBar, manifest []byte, exports []string, opts *osbuild.OSBuildOptions) (err error) {
 	rp, wp, err := os.Pipe()
 	if err != nil {
 		return fmt.Errorf("cannot create pipe for osbuild: %w", err)
 	}
 	defer rp.Close()
 	defer wp.Close()
 
-	cmd := newOsbuildCmd(manifest, exports, opts)
-	cmd.Args = append(cmd.Args, "--monitor=JSONSeqMonitor")
-	cmd.Args = append(cmd.Args, "--monitor-fd=3")
 
+	opts.Monitor = osbuild.MonitorJSONSeq
+	opts.MonitorFD = 3
+	opts.MonitorFile = wp
 	var stdio bytes.Buffer
-	var mw, buildLog io.Writer
-	var writeMu sync.Mutex
-	if opts.BuildLog != nil {
-		mw = newSyncedWriter(&writeMu, io.MultiWriter(&stdio, opts.BuildLog))
-		buildLog = newSyncedWriter(&writeMu, opts.BuildLog)
-	} else {
-		mw = &stdio
+	var mu sync.Mutex
+	buildLog := opts.BuildLog
+	opts.BuildLogMu = &mu
```

Comment:
Can someone explain why `newSyncedWriter ` was written in a way it accepts a mutex? I do not understand why a mutex could be shared for this decorator design pattern. It should have been:

```go
type SyncedWriter struct {
	mu sync.Mutex
	w  io.Writer
}

func NewSyncedWriter(w io.Writer) *SyncedWriter {
	return &SyncedWriter{w: w}
}

func (sw *SyncedWriter) Write(p []byte) (n int, err error) {
	sw.mu.Lock()
	defer sw.mu.Unlock()

	return sw.w.Write(p)
}
```

Than we could drop this passing of mutex, it is created on stack either way.

Reply:
Hmmm, right. Well, how about to tee before sync so each writer has its individual mutex that can be shared? Something like:

```go
syncedStdio := newSyncedWriter(&stdio)
syncedBuildLog := newSyncedWriter(opts.BuildLog)
io.MultiWriter(syncedStdio, syncedBuildLog)
```

Then use `syncedBuildLog` to write separately only to the build log independent of stdio. The example assumes `newSyncedWriter` no longer takes a shared mutex.

Reply:
<img width="1270" height="948" alt="image" src="https://github.com/user-attachments/assets/724fe20a-53fd-4742-a9f7-365a79874909" />
maybe helpful